### PR TITLE
Remove two redundant parenthesizations

### PR DESCRIPTION
**Pull request by shayne-fletcher-da**
Two redundant parenthesizations removed from the record processor. Also, remove redundant `$` operators from three tests.

### CODE REVIEW

File: compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/Records.hs [hunk 1]
```diff
@@ -158,13 +158,14 @@ onExp (L o (SectionR _ mid@(isDot -> True) rhs))
...
-    = setL o $ mkParen $ foldl1 (\x y -> noL $ OpApp noE x (mkVar var_dot) y) $ map (mkVar var_getField `mkAppType`) $ reverse sels
+    -- Don't bracket : it came in as a section so it's already enclosed.
+    = setL o $ foldl1 (\x y -> noL $ OpApp noE x (mkVar var_dot) y) $ map (mkVar var_getField `mkAppType`) $ reverse sels
 
 -- Turn a{b=c, ...} into setField calls
 onExp (L o upd@RecordUpd{rupd_expr,rupd_flds=L _ (HsRecField (fmap rdrNameAmbiguousFieldOcc -> lbl) arg pun):flds})
     | let sel = mkSelector lbl
     , let arg2 = if pun then noL $ HsVar noE lbl else arg
-    , let expr = mkParen $ mkVar var_setField `mkAppType` sel `mkApp` arg2 `mkApp` mkParen rupd_expr
+    , let expr = mkParen $ mkVar var_setField `mkAppType` sel `mkApp` arg2 `mkApp` rupd_expr -- 'rupd_expr' never appears to need bracketing.
```
Comment (cocreature) [inline compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/Records.hs]: “never appears to need bracketing” seems a bit too unsure to make me comfortable :slightly_smiling_face:. I don’t really know how GHC resolves precedences but it seems like you could end up with `(setField @"foo" x y) + z` instead of `setField @"foo" x (y + z)`. Have you tested this and similar cases, e.g., it would also be interesting to have another function application there so `y z`.

File: compiler/damlc/tests/daml-test-files/Choice_no_params_template.daml [hunk 2]
```diff
@@ -22,8 +22,8 @@ template Bond
       Split : (ContractId Bond, ContractId Bond)
         with splitAmount : Decimal
         do
-          f <- create $ this with amount = splitAmount
-          s <- create $ this with amount = amount - splitAmount
+          f <- create this with amount = splitAmount
```
Comment (cocreature) [inline compiler/damlc/tests/daml-test-files/Choice_no_params_template.daml]: Does this change require removing the `$`s or is this an unrelated change because they are redundant?

File: compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/Records.hs (hunk 1 above)
Comment (cocreature) [reply compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/Records.hs]: We probably at least want to have some test cases that hit this.

File: compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/Records.hs [hunk 3]
```diff
@@ -158,13 +158,14 @@ onExp (L o (SectionR _ mid@(isDot -> True) rhs))
     , srcSpanStart o == srcSpanStart (getLoc mid)
     , srcSpanEnd o == srcSpanEnd (getLoc rhs)
     , Just sels <- getSelectors rhs
-    = setL o $ mkParen $ foldl1 (\x y -> noL $ OpApp noE x (mkVar var_dot) y) $ map (mkVar var_getField `mkAppType`) $ reverse sels
+    -- Don't bracket : it came in as a section so it's already enclosed.
```
Comment (cocreature) [inline compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/Records.hs]: This is not about `:` is it? At least according to the comment above this is about sections on record accessors so `(.foo.bar)`. It is also not clear to me how this ends up looking in the AST. Do we end up with `mkParen (SectionR …)` or does the whole thing parse as `SectionR …`? In the latter case, removing the parenthesis is not going to work.

**Comment by hurryabit** [comment]
Can we please get a test that we can get something like the following through linting:
```
data A = A with x: Int
data B = B with y: A; z: A

f : Int -> A -> B
f x a = B (a with x) (a with x)
```

File: compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/Records.hs (hunk 1 above)
Comment (shayne-fletcher-da) [reply compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/Records.hs]: `rupd_expr` is the record part of the update in `RecordUpd {rupd_expr; rupd_fields}`. I think there are 3 possibilities : `rupd_expr` is a var, `rupd_expr` is an application, `rupd_expr` is a record constructor/update. (1) doesn't need bracketing, (2) needs to be in brackets already or it won't parse and (3) never needs bracketing.

File: compiler/damlc/tests/daml-test-files/Choice_no_params_template.daml (hunk 2 above)
Comment (shayne-fletcher-da) [reply compiler/damlc/tests/daml-test-files/Choice_no_params_template.daml]: This is unrelated, it just so happened I found them while testing. They are redundant.

File: compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/Records.hs (hunk 3 above)
Comment (shayne-fletcher-da) [reply compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/Records.hs]: This is about sections. It is lexically impossible to write a "raw" section in Haskell - they must be parenthesized. That means that the expression being transformed is in the context of an enclosing brackets. Bracketing here the causes double ups `((expr))`.

Comment (shayne-fletcher-da) [reply compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/Records.hs]: To clarify, `(.foo.bar)` is parsed as `HsPar (HsSectionR {...})`.

Comment (shayne-fletcher-da) [reply compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/Records.hs]: I'll improve the comment.

**Comment by shayne-fletcher-da** [comment]
Test program:
```
data A = A with x: Int
data B = B with y: A; z: A

f : Int -> A -> B
f x a = B (a with x) (a with x)
```
Linter:
```
INFO: Build completed successfully, 1 total action
No hints
```

**Comment by shayne-fletcher-da** [comment]
> Approving since I assume that you tested this manually but I’m somewhat unsure about both cases 

I've tested this against the DAML language tests. Together with my explanations above does that improve confidence a little?

> so it would be great if we could get a test file that tests all these edge cases and makes sure they are parsed properly.

I can spend some time writing more tests. I'd be grateful (as @hurryabit did just now) for any test cases you can think of.

**Comment by hurryabit** [comment]
@shayne-fletcher-da Can we please add the test to `compiler/damlc/tests/damlc-test-files`.

File: compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/Records.hs (hunk 1 above)
Comment (shayne-fletcher-da) [reply compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/Records.hs]: I'll fix the comment. I agree it doesn't inspire confidence :) I wrote it based on practice before I'd convinced myself by theory.

Comment (shayne-fletcher-da) [reply compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/Records.hs]: (1) `-- 'rupd_expr' never needs bracketing.`
(2) ``` -- Don't bracket here. The argument came in as a section so it's
    -- already enclosed in brackets.```

**Comment by shayne-fletcher-da** [comment]
> @shayne-fletcher-da Can we please add the test to `compiler/damlc/tests/damlc-test-files`.

Yes-sir.

File: compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/Records.hs (hunk 1 above)
Comment (cocreature) [reply compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/Records.hs]: I’m still not convinced, sorry :slightly_smiling_face:  In particular, I am not convinced that 2 holds. If I have `x { a = y, b = z }` I believe you will first rewrite this to `setField @"a" x { b = z }` and then when you rewrite this it breaks.

**Comment by shayne-fletcher-da** [comment]
Actually, @martin-drhu-da : `f x a = B (a with x) (a with x)` has a lint.
```
Message: 
  /Users/shaynefletcher/project/daml.git/compiler/damlc/tests/daml-test-files/RecordUpdateExp.daml:13:11:
  Warning: Redundant bracket
  Found:
  ((DA.Internal.Record.setField @"x" x a))
  Perhaps:
  (DA.Internal.Record.setField @"x" x a)
File:    
  /Users/shaynefletcher/project/daml.git/compiler/damlc/tests/daml-test-files/RecordUpdateExp.daml
Range:    13:21-13:31
Source:   linter
Severity: DsInfo
Message: 
  /Users/shaynefletcher/project/daml.git/compiler/damlc/tests/daml-test-files/RecordUpdateExp.daml:13:22:
  Warning: Redundant bracket
  Found:
  ((DA.Internal.Record.setField @"x" x a))
  Perhaps:
  (DA.Internal.Record.setField @"x" x a)
```
Removing the parentheses fixes the lint.

File: compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/Records.hs (hunk 1 above)
Comment (shayne-fletcher-da) [reply compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/Records.hs]: In the above `x` is the `rupd_expr`.

Comment (shayne-fletcher-da) [reply compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/Records.hs]: Test:
```
data C = C {a : Int, b: Int}
g x y z = x {a = y, b = z}
```
Compiles and is lint free.

Comment (shayne-fletcher-da) [reply compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/Records.hs]: Which 2 is it that you are not convinced holds?

Comment (shayne-fletcher-da) [reply compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/Records.hs]: > “never appears to need bracketing” seems a bit too unsure to make me comfortable 🙂. I don’t really know how GHC resolves precedences but it seems like you could end up with `(setField @"foo" x y) + z` instead of `setField @"foo" x (y + z)`. Have you tested this and similar cases, e.g., it would also be interesting to have another function application there so `y z`.

Here's some tests that I think covers those?
```
data C = C {a : Int, b: Int}
g x y z = x {a = y, b = z}
h x y z = x {a = y + z}

mkC a b = C with a, b

main = scenario do
  let x = C {a = 1, b = 2}
  let x' = g x 3 4
  assert $ x'.a == 3 && x'.b == 4

  let x'' = h x 1 2
  assert $ x''.a == 3

  let x''' = h (mkC 5 6) 1 1
  assert $ x'''.a == 2
```

Comment (shayne-fletcher-da) [reply compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/Records.hs]: One more from above you asked for and I missed `i x y =  x {a = succ y}` :
```
  let x'''' = i C {a = 1, b = 2} 2
  assert $ x''''.a == 3
```

Comment (cocreature) [reply compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/Records.hs]: If it works then all is good but I’m still not sure I understand how it works. Is it as simple as GHC doesn’t care if `rupd_expr` has parentheses or not? It seems like you will definitely not always have parentheses around the expression if there are multiple field updates.

**Comment by shayne-fletcher-da** [comment]
> I can spend some time writing more tests.

I think we're all beginning to believe? In any case, this is scary enough (for me included) that I'll commit a big-ass test along with it incorporating Dr. Hu's test above, all the cases inlined in this ticket and anything else I can think of.

File: compiler/damlc/tests/daml-test-files/RecordUpdateExp.daml [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+daml 1.2
+module RecordUpdateExp where
+
+data A = A with x: Int
+data B = B with y: A; z: A
+
+-- Check parsing/linting.
+
+f : Int -> A -> B
+f x a = B a with
```
Comment (hurryabit) [inline compiler/damlc/tests/daml-test-files/RecordUpdateExp.daml]: Is there a way to put this onto one line without getting the "redundant brackets" lint? If not, we should disable the lint by default.

Comment (shayne-fletcher-da) [reply compiler/damlc/tests/daml-test-files/RecordUpdateExp.daml]: Oops that was my mistake mucking around. Let me fix that.

Comment (shayne-fletcher-da) [reply compiler/damlc/tests/daml-test-files/RecordUpdateExp.daml]: Fixed. As I mentioned earlier `f x a = B a with x a with x` is perfectly valid.
